### ISSUE TITLE
Class-Transformer Collection Issue

### ISSUE THREAD

**Issue by hasezoey** [typegoose/typegoose#108]
Currently `class-transformer` dosnt work on documents, because it is ***not*** an instance of the class, and without an function like `anyToplain(class, input)` (https://github.com/typestack/class-transformer/issues/227), it is currently not possible to use `class-transformer` (or `class-validator`) with typegoose

-> when using it with nestjs and returning it in an route, you should always call `doc.toJSON()` (or directly `model.find().lean()` to directly convert it to a POJO

## Related Issues

- #61 
- #96 
(probably - #105 too)

**Comment by sebastiangug** [comment]
Hi not sure if I'm confusing things -- do you mean if a key on a document has the value of another class/subcollection with ref? or just class-transformer not working with any use case? 

I'm currently using class-transformer/class-validator with my classes but my data models do not need subcollections/subdocuments or use "ref" as part of a nestjs based project. 

I haven't encountered any issues with it.

**Comment by hasezoey** [comment]
@sebastiangug in my testing (back in 5.x and 6.x times), class-transformer didnt want to translate any document correctly
-> i never tested it on 7.x, because from what i know, the code on how typegoose and mongoose handles this is the same

**Comment by hasezoey** [comment]
i dont know if it will work now, but if anyone knows how to use class-transformer / class-validator, i would like to accept pr's that add tests for them

**Comment by Roms1383** [comment]
As a side note `class-validator` is similar to `class-transformer` in the way that it can only work its magic after creating an instance of its annotated class. I'm not sure that showing its integration with `typegoose` is really relevant, unless you want to validate the result of a `typegoose` query, since usually it's used upfront to validate data arriving on your route endpoints.

**Comment by hasezoey** [comment]
@Roms1383 from what i know c-v is commonly used with type-graphql and nestjs

**Comment by Roms1383** [comment]
Yes for Nest.js [as specified in the documentation](https://docs.nestjs.com/pipes#class-validator) validation usually happens on your route controller.

For incoming data send to the `Controller` route, `ValidationPipe` with option `transform: true` will both deserialize incoming plain payload with `class-transformer` into a class instance and validate it with `class-validator` altogether.

For outgoing data returned from `Controller` route, usually serialization from a class instance back to a plain object happens with `ClassSerializerInterceptor` specified on the `Controller` route.

e.g.
```ts
// user.controller.ts
@Controller()
export class UserController {
  constructor(
     private readonly users: UserService,
     private readonly mailer: MailerService) {}

  @UseInterceptors(ClassSerializerInterceptor) //serialize outgoing data
  @Post('register')
  register(
    @Body(new ValidationPipe({ transform: true })) // deserialize & validate incoming data
    payload: User) {
    return this.users.create(payload)
  }
}
```
```ts
// user.service.ts
@Injectable()
export class UserService {
  private model: any
  constructor() {
    this.model = getModelForClass(User)
  }
  async create(user: User): Promise<Document<User>|never> {
    const { _id: id } = await this.model.create(user)
    const newcomer = await this.model.findById(id).exec()
    return newcomer
  }
}
```

When it comes to `typegoose`, it might be convenient to deserialize database queries back to class instances if there's additional logic happening inside the backend.

e.g.
```ts
// user.controller.ts
@Controller()
export class UserController {
  constructor(
     private readonly users: UserService,
     private readonly mailer: MailerService) {}

  @UseInterceptors(ClassSerializerInterceptor) //serialize outgoing data
  @Post('register')
  register(
    @Body(new ValidationPipe({ transform: true })) // deserialize & validate incoming data
    payload: User) {
    const newcomer: User = await this.users.create(payload)
    this.mailer.greet(newcomer) // additional logic
    return newcomer
  }
}
```
```ts
// user.service.ts
@Injectable()
export class UserService {
  private model: any
  constructor() {
    this.model = getModelForClass(User)
  }
  async create(user: User): Promise<User|never> {
    const { _id: id } = await this.model.create(user)
    const newcomer = await this.model.findById(id).exec()
    return plainToClass(User, newcomer) // deserialize plain Document<User> back into a User instance
  }
}
```

I'm not familiar with GraphQL, so if somebody who does can step in and had more informations it would be nice 😃

**Comment by hasezoey** [comment]
This should be fixed since #402, and with 116599cb0b8589d2826a2de70a484aba5716de12 updated to ensure being compatible with the latest version of `class-transformer` (which currently is `0.4.0`)

**Comment by smolinari** [comment]
Just to add the note that with GraphQL resolvers (Queries/ Mutations), it is basically the same situation (with NestJS). I'd suggest though that whether Controller, Mutation or Query, the transformations should be happening just above the "Service" level, where the specific business logic is brought together with the data access layer (Typegoose). Nest has the Service concept built in and has things mentioned like Pipes and Interceptors. With TypeGraphQL, you'd need to add some sort of modules system to get a similar architecture going. 

I guess the point I'm trying to make is, the transformation logic should be in a layer between Typegoose and the rest of the application and usually you'll have business logic before that extra layer too. 

As a perfect example, I'm working with a package called Nestjs-Query and it has "transformation" built in above its Services offering. They are called Assemblers and are pretty cool. 

https://doug-martin.github.io/nestjs-query/docs/concepts/advanced/assemblers

Scott
